kubeadm: change service activity check logic in ServiceCheck

mythi opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
The PR brings kubeadm improvements.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
The PR changes the service activity check logic in ServiceCheck. It is necessary because kubelet may be running (in the standalone mode) and the user doesn't necessarily now why the other checks (especially Port10250) are failing.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
kubeadm: change service activity check logic in ServiceCheck 
```

mythi commented:
cnfc-cla should be OK now.

timothysc commented on cmd/kubeadm/app/preflight/checks.go:
```diff
@@ -148,10 +148,16 @@ func (sc ServiceCheck) Check() (warnings, errors []error) {
 
-	if sc.CheckIfActive && !initSystem.ServiceIsActive(sc.Service) {
+	if sc.WantActive != initSystem.ServiceIsActive(sc.Service) {
+		status  := [2]string{"active", "not active"}
```
I'm guessing some tests will need to be updated for this?

mythi replied:
```make check WHAT="./cmd/kubeadm/app/preflight" GOFLAGS="-v"``` passes with these changes. I'll need to see why the automated PR checks are failing.

xiangpengzhao commented:
/ok-to-test

@timothysc FYI: removing `needs-ok-to-test` manually doesn't trigger tests :)

xiangpengzhao commented:
@mythi seems like CLA status is not detected...

xiangpengzhao commented:
/retest
Seems like they are test flakes.

xiangpengzhao commented:
k8s-ci-robot added the cncf-cla: no  label 12 hours ago

@mythi your CLA status seems still not correct :)

xiangpengzhao commented:
```
      diff -u ./cmd/kubeadm/app/preflight/checks.go.orig ./cmd/kubeadm/app/preflight/checks.go
--- ./cmd/kubeadm/app/preflight/checks.go.orig	2018-02-24 03:22:20.945495368 +0000
+++ ./cmd/kubeadm/app/preflight/checks.go	2018-02-24 03:22:20.945495368 +0000
@@ -115,9 +115,9 @@
 // detect a supported init system however, all checks are skipped and a warning is
 // returned.
 type ServiceCheck struct {
-	Service       string
+	Service    string
 	WantActive bool
-	Label         string
+	Label      string
 }
 
 // Name returns label for ServiceCheck. If not provided, will return based on the service parameter
@@ -149,10 +149,10 @@
 	}
 
 	if sc.WantActive != initSystem.ServiceIsActive(sc.Service) {
-		status  := [2]string{"active", "not active"}
+		status := [2]string{"active", "not active"}
 		command := [2]string{"stop", "start"}
-		index   := 0
-		if (sc.WantActive) {
+		index := 0
+		if sc.WantActive {
 			index = 1
 		}
 		errors = append(errors,

Run ./hack/update-gofmt.sh
```

mythi commented:
On Tue, Feb 27, 2018 at 9:08 AM, Peter (XiangPeng) Zhao wrote:

> k8s-ci-robot added the cncf-cla: no label 12 hours ago
>
> @mythi <https://github.com/mythi> your CLA status seems still not correct
> :)
>
I noticed this and I'm currently checking why the robot thinks it's not OK.
I have it signed.

mythi commented:
/test cla/linuxfoundation

mythi commented:
/retest

mythi commented:
@xiangpengzhao how do I get the test re-triggered? I'd want to check if the bot is able see my CLA status.

xiangpengzhao commented:
IIRC, the CLA status check will be triggered when there is new comments on the PR. @cjwagner has the CLA plugin behavior changed recently ?

mythi commented:
On Thu, Mar 1, 2018 at 4:32 PM, Peter (XiangPeng) Zhao wrote:

> IIRC, the CLA status check will be triggered when there is new comments on
> the PR. @cjwagner <https://github.com/cjwagner> has the CLA plugin
> behavior changed recently ?
>

I've tried all possible triggers (re-push, /(re)test etc.) :) Perhaps the
label needs to be removed to get it re-triggered.

xiangpengzhao commented:
@mythi can you create a new PR (whatever the it's commit is, just for testing) to see if CLA can be detected in the new PR? If it can be detected, you can ask @timothysc to label `cla:human-approved` for you.

cjwagner commented:
The CLA Prow plugin just manages the `cncf-cla: yes` and `cncf-cla: no` labels based on the "cla/linuxfoundation" status context which is controlled by the CNCF, not Prow.

Pushing a new commit may retrigger the CNCF job that sets the "cla/linuxfoundation" status?

mythi commented:
On Thu, Mar 1, 2018 at 8:38 PM, Cole Wagner wrote:

> The CLA Prow plugin just manages the cncf-cla: yes and cncf-cla: no
> labels based on the "cla/linuxfoundation" status context which is
> controlled by the CNCF, not Prow.
>
> Pushing a new commit may retrigger the CNCF job that sets the
> "cla/linuxfoundation" status?
>
>
Neither new commit nor close/reopen cycle didn't re-trigger the check.
Before trying a new PR, perhaps someone could remove the cncf-cla label
here to see if that makes the difference.

BenTheElder commented:
It's not the CLA label, the `cla/linuxfoundation` status only inspects git commit emails, which is independent of the label being added by our bot once that status changes. Is it possible one of your commits has the wrong email or the CLA is not signed with your account yet?

mythi commented:
Thanks for the clarification! I've double checked the email in my commits are correct and the CLA is ok.

BenTheElder commented:
@mythi Sorry I can't do much more here :/
You should try contacting the linux foundation help desk next, see: https://github.com/kubernetes/kubernetes/issues/27796#issuecomment-313159219

mythi commented:
@BenTheElder No problem, I'm currently checking this with LF helpdesk

BenTheElder commented:
once https://github.com/kubernetes/kubernetes/pull/60732 merges we can retest verify

thockin commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/60304)
<!-- Reviewable:end -->

xiangpengzhao commented:
/retes

mythi commented:
I pushed v2 yesterday and dropped/moved the kernel config related commit to its own PR (#60791)

mythi commented:
re-testing cla/linuxfoundation

BenTheElder commented:
/test pull-kubernetes-typecheck

BenTheElder commented:
ah, the import change needs a `hack/update-bazel.sh`.

mythi commented:
On Thu, Mar 8, 2018 at 8:32 AM, Benjamin Elder <notifications@github.com>
wrote:

> ah, the import change needs a hack/update-bazel.sh.
>
>
okay, thanks. It currently fails for me so I'll need to see what's wrong
with that before I can update.

```
!!! Error in hack/update-bazel.sh:51
  Error in hack/update-bazel.sh:51. 'kazel' exited with status 255
Call stack:
  1: hack/update-bazel.sh:51 main(...)
Exiting with status 1
```

BenTheElder commented:
Hmm is that the entire output? The last person I talked to with kazel failing just had kubernetes checked out to the wrong location (should be `$GOPATH/src/k8s.io/kubernetes`)

mythi commented:
@BenTheElder That was it!

mythi commented:
/cc @kad

mythi commented:
retriggering cla check

xiangpengzhao commented:
>retriggering cla check

It works at last! :100:

timothysc commented on cmd/kubeadm/app/preflight/checks.go:
```diff
@@ -148,10 +148,16 @@ func (sc ServiceCheck) Check() (warnings, errors []error) {
-	if sc.CheckIfActive && !initSystem.ServiceIsActive(sc.Service) {
+	if sc.WantActive != initSystem.ServiceIsActive(sc.Service) {
+		status := [2]string{"active", "not active"}
+		command := [2]string{"stop", "start"}
```
This parsing of intent from the return is odd to me.  I'd think that having a utility function, or augmenting the return arg would be cleaner.

mythi replied:
I've changed this.

timothysc reviewed (changes requested):
Also please squash commits.

dims commented:
/uncc @dims

mythi committed:
kubeadm: preflight: change kubelet startup logic

Currently, the service activity checks logic is either

  a) I care about the active status and it needs to be active or
  b) I don't care about the active status

This limits the usage and won't allow to check

  c) I care about the status and it needs to be NOT active

The additional check is necessary to be control kubelet is started
right time in the bootstrap process. The change in the preflight checks
now errors if kubelet is running when kubeadm is called.

Additionally, we change to start kubelet later in the bootstrap
process. Currently, kubelet is started right after the preflight
checks but before the configs are created. At that point, kubeadm
has the assumption that kubelet goes in a crash-loop and starts OK
after the configs are created.

It may be that the crash-loop does not happen but kubelet runs
in a standalone mode (e.g., in case if kubelet is started without
--bootstrap-kubeconfig). This makes kubeadm to fail due to a
timeout in markmaster when running 'kubeadm init'.

mythi commented:
rebased + addressed the review feedback.

timothysc commented:
/assign @detiber @stealthybox

detiber commented on cmd/kubeadm/app/preflight/checks.go:
```diff
@@ -987,7 +990,7 @@ func addCommonChecks(execer utilsexec.Interface, cfg kubeadmapi.CommonConfigurat
 		HostnameCheck{nodeName: cfg.GetNodeName()},
 		KubeletVersionCheck{KubernetesVersion: cfg.GetKubernetesVersion(), exec: execer},
-		ServiceCheck{Service: "kubelet", CheckIfActive: false},
+		ServiceCheck{Service: "kubelet", WantActive: false},
```
This seems to change the behavior of the kubelet check significantly here...  It looks like previously CheckIfActive would skip the active check completely, but now with WantActive this is testing specifically for the kublet not being active.

mythi replied:
Yes, that's what I'm suggesting. The reasons are explained in the commit message.

detiber replied:
Right, but since we are checking for `active` or `activating` [here](https://github.com/kubernetes/kubernetes/blob/7b8bb6e7d3345b04678d96a721f651a4a68eb1dd/pkg/util/initsystem/initsystem.go#L78) wouldn't this preflight check fail by default on any system that has the kubelet enabled/started before running kubeadm (which is the default case on Ubuntu/Debian systems)?

mythi replied:
Here's my observation: if kubelet is ```active```, ```kubeadm init``` would fail in markmaster timeout (I was hit by this). If kubelet is ```activating```, it's actually crash-looping and can only start/recover once the configurations are created by ```kubeadm```. 

So I thought to make it more bullet proof and make sure we start with kubelet not running at all. That's the state where ```kubeadm reset``` leaves it too.

detiber replied:
@mythi that behavior makes perfect sense to me and is the default behavior when installing from the rpms. However the expected behavior with debs is that the service is enabled and started after installation, which is also the behavior with the current debs that we ship.

While I personally don't agree with the default behavior of enabling/starting daemons on package install, it is the expected behavior on Debian/Ubuntu systems.

That said, I believe the only reason that we are currently checking for `active` or `activating` is for the check in kubeadm.

mythi replied:
@detiber I'm not sure I got your point wrt. to my change proposal. I don't mind to keep kubelet running as it is today (and I don't mind if this PR is not OK) as long as the behaviour is consistent. Right now you have to have it _crash-looping_ in order to get ```kubeadm init``` to succeed. For newcomers like me it'll take some time to realize why ```active``` is not sufficient.

detiber replied:
@mythi maybe I'm not fully grokking these changes, but it appears to me that the preflight check would currently fail on an Ubuntu host because we are testing that initSystem.ServiceIsActive() returns false for the kubelet, since ServiceIsActive() is checking for 'active' or 'activating' and the kubelet service on Ubuntu will be in 'activating' by default.

mythi replied:
@detiber Currently the preflight check for kubelet ignores the activity status (```CheckIfActive``` is set to ```false```).

mythi commented:
I don't have the urgent need to get this merged since I know the workaround needed to address the failures I'm seeing. I'm closing this and look into alternatives to improve documentation to ensure people don't hit the same problem.
